commands/src/main/java/org/wildfly/extras/creaper/commands/modules/AddModule.java

```diff
@@ -0,0 +1,171 @@
+import org.wildfly.extras.creaper.core.online.OnlineCommand;
+import org.wildfly.extras.creaper.core.online.OnlineCommandContext;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * <p>The command is used to add modules</p>
+ *
+ * <p>The corresponding to the module name directory structure will be created in the AS module repository.
+ * The JAR files specified as resources will be copied to the module's directory. Unless module.xml file
+ * was specified as an argument, it will be automatically generated.</p>
+ *
+ * <p>Example of resulting cli command:
+ * module add --name=name --slot=slot --resources=jars
+ * --resource-delimiter=char --dependencies=modules --properties=properties</p>
+ */
+public class AddModule implements OnlineCommand {
+    private final String resourceDelimiter;
+    private final String moduleName;
+    private final String slot;
+    private final String mainClass;
+    private final String moduleXml;
+    private final List<String> resources;
+    private final List<String> dependencies;
+    private final List<String> properties;
+
+    private AddModule(Builder builder) {
+        this.resourceDelimiter = builder.resourceDelimiter;
+        this.moduleName = builder.moduleName;
+        this.slot = builder.slot;
+        this.mainClass = builder.mainClass;
+        this.moduleXml = builder.moduleXml;
+        this.resources = builder.resources;
+        this.dependencies = builder.dependencies;
+        this.properties = builder.properties;
+    }
+
+    @Override
+    public void apply(OnlineCommandContext ctx) throws Exception {
+        if (ctx.options.isDomain) {
+            throw new CommandFailedException("AddModule command isn't supported in domain mode");
+        }
+
+        StringBuilder cmd = new StringBuilder("module add");
+        cmd.append(" --name=" + moduleName);
+        cmd.append(" --slot=" + slot);
+        cmd.append(" --resource-delimiter=" + resourceDelimiter);
+
+        if (moduleXml != null)
+            cmd.append(" --module-xml=" + moduleXml);
+
+        if (mainClass != null)
+            cmd.append(" --main-class=" + mainClass);
+
+        Joiner resourcesJoiner = Joiner.on(resourceDelimiter);
+        if (!resources.isEmpty()) {
+            cmd.append(" --resources=" + resourcesJoiner.join(resources).replaceAll(" ", "\\\\ "));
+        }
+
+        Joiner defaultJoiner = Joiner.on(",");
+        if (!dependencies.isEmpty()) {
+            cmd.append(" --dependencies=" + defaultJoiner.join(dependencies));
+        }
+
+        if (!properties.isEmpty()) {
+            cmd.append(" --properties=" + defaultJoiner.join(properties));
+        }
+        ctx.client.executeCli(cmd.toString());
+    }
+
+    @Override
+    public String toString() {
+        return "AddModule " + moduleName;
+    }
+
+    public static final class Builder {
+        public static final String DEFAULT_SLOT = "main";
```

Comment:
I'd probably move this to the top-level `AddModule` class, because I think that a builder shouldn't hold constants related to the class it builds, but I'll leave that up to you.

I was thinking about making it `private`/`protected`/default, but I agree that it makes sense as `public`.